source/server/listener_impl.cc

```diff
@@ -59,42 +94,30 @@ Network::SocketSharedPtr ListenSocketFactoryImplBase::createListenSocketAndApply
   return socket;
 }
 
-void ListenSocketFactoryImplBase::setLocalAddress(
-    Network::Address::InstanceConstSharedPtr local_address) {
-  ENVOY_LOG(debug, "Set listener {} socket factory local address to {}", listener_name_,
-            local_address->asString());
-  local_address_ = local_address;
-}
-
-TcpListenSocketFactory::TcpListenSocketFactory(
-    ListenerComponentFactory& factory, Network::Address::InstanceConstSharedPtr local_address,
-    const Network::Socket::OptionsSharedPtr& options, bool bind_to_port,
-    const std::string& listener_name)
-    : ListenSocketFactoryImplBase(factory, local_address, Network::Address::SocketType::Stream,
-                                  options, bind_to_port, listener_name) {
-  socket_ = createListenSocketAndApplyOptions();
-  if (socket_ != nullptr && localAddress()->ip() != nullptr && localAddress()->ip()->port() == 0) {
-    setLocalAddress(socket_->localAddress());
+Network::SocketSharedPtr ListenSocketFactoryImpl::getListenSocket() {
+  if (!reuse_port_) {
+    return socket_;
   }
-}
 
-Network::SocketSharedPtr TcpListenSocketFactory::getListenSocket() { return socket_; }
-
-UdpListenSocketFactory::UdpListenSocketFactory(
-    ListenerComponentFactory& factory, Network::Address::InstanceConstSharedPtr local_address,
-    const Network::Socket::OptionsSharedPtr& options, bool bind_to_port,
-    const std::string& listener_name)
-    : ListenSocketFactoryImplBase(factory, local_address, Network::Address::SocketType::Datagram,
-                                  options, bind_to_port, listener_name) {}
-
-Network::SocketSharedPtr UdpListenSocketFactory::getListenSocket() {
-  // TODO(danzh) add support of SO_REUSEPORT. Currently calling this method twice will fail because
-  // the port is already in use.
-  Network::SocketSharedPtr socket = createListenSocketAndApplyOptions();
-  if (socket != nullptr && localAddress()->ip() != nullptr && localAddress()->ip()->port() == 0) {
-    setLocalAddress(socket->localAddress());
+  Network::SocketSharedPtr socket;
+  absl::call_once(steal_once_, [this, &socket]() {
+    if (socket_) {
+      // If a listener's port is set to 0, socket_ should be created for reserving a port
+      // number, it is handed over to the first worker thread came here.
+      // There are several reasons for doing this:
+      // - for UDP, once a socket being bound, it begins to receive packets, it can't be
```

Comment:
As an aside, I'm not sure how much this really matters. There is going to be a delay between creating the socket on the main thread and when the first worker actually steals it (the delay may be very large due to xDS init, etc.). In general this should be covered by downstream health check, etc. This is related to an offline conversation I had with @danzh2010 about UDP and hot restart. I think it probably will never work correctly with stateless protocols like QUIC. I will open an issue to figure out how to deal with that.